url0e7d57473a_kyorohiro_HetimaUtil_tgz-pJ8-net_hetimatan_net_http_HttpServer3xxJ8/src/net/hetimatan/io/file/TestForVirtualFile.java

```diff
@@ -0,0 +1,319 @@
+			assertTrue(false);
+		} finally {
+			if(vf != null) {
+				try {
+					vf.close();
+				} catch (IOException e) {
+					e.printStackTrace();
+				}
+			}
+		}
+	}
+
+
+	public void testThird() {
+		File testPath = new File(getFile(),"__001__.txt");
+		String testdata= ""
+						+"「ウップ。怪しい結論だね。恐ろしく無駄骨の折れる虚栄じゃないか」\r\n"
+						+"「ええ。それがね。あの人は地道に行きたい行きたい。"
+						+"みんなに信用されていたいいたいと、思い詰めているのがあの娘"
+						+"ひと"
+						+"の虚栄なんですからね。そのために虚構"
+						+"うそ"
+						+"を吐"
+						+"つ"
+						+"くんですよ」\r\n"
+						+"「それが第一おかしいじゃないか。第一、そんなにまでしてこちらの信用を博する必要が何処に在るんだい。看護婦としての手腕はチャント認められているんだし、実家"
+						+"うち"
+						+"が裕福だろうが貧乏だろうが看護婦としての資格や信用には無関係だろう。それくらいの事がわからない馬鹿じゃ、姫草はないと思うんだが」\r\n"
+						+"「ええ。そりゃあ解ってるわ。たとえドンナ女"
+						+"ひと"
+						+"だっても現在ウチの病院の大切なマスコットなんですから、疑ったり何かしちゃすまないと思うんですけど……ですけど毎月二日か三日頃になると印形"
+						+"ハンコ"
+						+"で捺"
+						+"お"
+						+"したように白鷹先生の話が出て来るじゃないの。おかしいわ……」\r\n"
+						+"「そりゃあ庚戌会がその頃にあるからさ」\r\n";
+		CashKyoroFile vf = null;
+		try {
+			testPath.delete();
+			testPath.createNewFile();
+			vf = new CashKyoroFile(testPath, 1024, 3);
+			byte[] buffer = new byte[1300];
+			for(int i=0;i<buffer.length;i++) {
+				buffer[0] = 0;
+			}
+			int ret = vf.read(buffer);
+			assertEquals(-1, ret);
+			for(int i=0;i<buffer.length;i++) {
+				assertEquals(0, buffer[0]);
+			}
+			byte[] testdataBuffer = testdata.getBytes("utf8");
+			vf.addChunk(testdataBuffer, 0, testdataBuffer.length);
+			{
+				//
+				assertTrue(""+testdataBuffer.length +"<="+buffer.length,testdataBuffer.length <= buffer.length);
+				// first read
+				ret = vf.read(buffer);
+				String result = new String(buffer, 0 , ret);
+				assertEquals(testdata, result);
+
+				// second read
+				{
+					for(int i=0;i<buffer.length;i++) {
+						buffer[0]=0;
+					}
+					ret = vf.read(buffer);
+					assertEquals(-1, ret);					
+				}
+			}
+			{
+				// sync
+				for(int i=0;i<buffer.length;i++) {
+					buffer[0]=0;
+				}
+				vf.syncWrite();
+				vf.seek(0);
+				ret = vf.read(buffer);
+				String result = new String(buffer, 0 , ret);
+				assertEquals(testdata, result);
+			}
+		} catch (Exception e) {
+			e.printStackTrace();
+			assertTrue(false);
+		} finally {
+			if(vf != null) {
+				try {
+					vf.close();
+				} catch (IOException e) {
+					e.printStackTrace();
+				}
+			}
+		}
+	}
+
+
+	public void testExtra1() {
+		File testPath = new File(getFile(),"__001__.txt");
+		String[] testdata= {
+				"●..1/:::/storage/sdcard0\r\n",
+				"  2013/01/12 20:41:08:::/storage/sdcard0\r\n",
+				"  8192byte:::/storage/sdcard0:::HR\r\n",
+				"●keyword/:::/storage/sdcard0/.jota/keyword\r\n",
+				"  2012/12/14 23:21:43:::/storage/sdcard0/.jota/keyword\r\n",
+				"  4096byte:::/storage/sdcard0/.jota/keyword:::HR\r\n",
+				"●..2/:::/storage/sdcard0\r\n",
+				"  2013/01/12 20:41:08:::/storage/sdcard0\r\n",
+				"  8192byte:::/storage/sdcard0:::HR\r\n"
+		};
+		try {
+			testPath.delete();
+			testPath.createNewFile();
+			CashKyoroFile vf = new CashKyoroFile(testPath, 1024, 5);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code contains a resource that might not have closed properly. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *vf*. The resource is referenced at lines 271, 278, 290, 293, 294, 302, 303, 304. The resource is closed at line 308. There are other execution paths that do not contain closure statements, for example, when CashKyoroFile.read throws an exception.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).